autometa/common/external/diamond.py

```diff
@@ -28,204 +28,359 @@
+        The subtraction will only be successful if both the DiamondResult objects have the same qseqid
+
+        Parameters
+        ----------
+        other_hit : Dict 
+            Other DiamondResult object to subtract
+
+        Returns
+        -------
+        DiamondResult object
+            A sseqid dict as now been subtracted (removed) from the sseqids dict in the object
+        
+        Raises
+        ------
+        AssertionError
+            Query sequences are not equal
+        """
+        assert (
+            self == other_hit
+        ), f"qseqids do not match! {self.qseqid} & {other_hit.qseqid}"
+        for key in list(other_hit.sseqids.keys()):
+            try:
+                self.sseqids.pop(key)
+            except KeyError as err:
+                raise err
         return self
 
     def get_top_hit(self):
-        top_bitscore = float('-Inf')
+        """
+        Returns the subject sequene ID with the highest bitscore amongst all the subject sequence that hit a query
+        """
+        top_bitscore = float("-Inf")
         top_hit = None
-        for sseqid,attrs in self.sseqids.items():
-            if attrs.get('bitscore', float('-Inf')) >= top_bitscore:
-                top_bitscore = attrs.get('bitscore')
+        for sseqid, attrs in self.sseqids.items():
+            if attrs.get("bitscore", float("-Inf")) >= top_bitscore:
+                top_bitscore = attrs.get("bitscore")
                 top_hit = sseqid
         return top_hit
 
-def makedatabase(fasta, database, nproc=1):
-    cmd = f'diamond makedb --in {fasta} --db {database} -p {nproc}'
-    logger.debug(f'{cmd}')
-    with open(os.devnull, 'w') as stdout, open(os.devnull, 'w') as stderr:
-        retcode = subprocess.call(cmd, stdout=stdout, stderr=stderr, shell=True)
-    if retcode:
-        raise OSError(f'DiamondFailed:\nArgs:{proc.args}\nReturnCode:{proc.returncode}')
+
+def makedatabase(fasta, database, cpus=mp.cpu_count()):
+    """
+    Creates a database against which the query sequence would be blasted
+
+    Parameters
+    ----------
+    fasta : str
+        Path to fasta file whose database needs to be made
+        e.g. '<path/to/fasta/file>'
+    database : str
+        Path to the output diamond formatted database file
+        e.g. '<path/to/database/file>'
+    cpus : int, optional
+        Number of processors to be used. By default uses all the processors of the system 
+
+    Returns
+    -------
+    str
+        Path to diamond formatted database
+
+    Raises
+    ------
+    subprocess.CalledProcessError
+        Failed to create diamond formatted database
+    """
+    cmd = f"diamond makedb --in {fasta} --db {database} -p {cpus}"
+    logger.debug(f"{cmd}")
+    proc = subprocess.run(
+        cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True
+    )
+    try:
+        proc.check_returncode()
+    except subprocess.CalledProcessError as err:
+        raise err
```

Comment:
1. Since you returning the same error message that is being returned from `subprocess.CalledProcessError`, you could place this within `subprocess.run(check=True)`
2. You are running `subprocess.run` with `shell=True`, this has been [documented as being a security risk](https://docs.python.org/3/library/subprocess.html#security-considerations). So we should try to minimize this as much as possible.
```suggestion
    cmd = ["diamond", "makedb", "--in", fasta, "--db", database ,"-p", cpus]
    subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
    logger.debug(" ".join(cmd))
```